plugins/incremental-ingestion-backend/src/engine/IncrementalIngestionEngine.ts

```diff
@@ -0,0 +1,286 @@
+              );
+
+              const backoffLength = currentBackoff.as('milliseconds');
+              this.options.logger.error(error);
+
+              const truncatedError =
+                typeof error === 'string'
+                  ? error.substring(0, 700)
+                  : `${error}`;
+              this.options.logger.error(
+                `incremental-engine: Ingestion '${ingestionId}' threw an error during ingestion burst. Ingestion will backoff for ${currentBackoff.toHuman()} (${truncatedError})`,
+              );
+
+              await this.manager.setProviderBackoff(
+                ingestionId,
+                attempts,
+                error as Error,
+                backoffLength,
+              );
+            }
+          }
+          break;
+        case 'backoff':
+          if (Date.now() > nextActionAt) {
+            this.options.logger.info(
+              `incremental-engine: Ingestion '${ingestionId}' backoff complete, will attempt to resume`,
+            );
+            await this.manager.setProviderIngesting(ingestionId);
+          } else {
+            this.options.logger.info(
+              `incremental-engine: Ingestion '${ingestionId}' backoff continuing`,
+            );
+          }
+          break;
+        case 'cancel':
+          this.options.logger.info(
+            `incremental-engine: Ingestion '${ingestionId}' canceling, will restart`,
+          );
+          await this.manager.setProviderCanceled(ingestionId);
+          break;
+        default:
+          this.options.logger.error(
+            `incremental-engine: Ingestion '${ingestionId}' received unknown action '${nextAction}'`,
+          );
+      }
+    } else {
+      this.options.logger.error(
+        `incremental-engine: Engine tried to create duplicate ingestion record for provider '${this.options.provider.getProviderName()}'.`,
+      );
+    }
+  }
+
+  async getCurrentAction() {
+    const providerName = this.options.provider.getProviderName();
+    const record = await this.manager.getCurrentIngestionRecord(providerName);
+    if (record) {
+      this.options.logger.info(
+        `incremental-engine: Ingestion record found: '${record.id}'`,
+      );
+      return {
+        ingestionId: record.id,
+        nextAction: record.next_action as 'rest' | 'ingest' | 'backoff',
+        attempts: record.attempts as number,
+        nextActionAt: record.next_action_at.valueOf() as number,
+      };
+    }
+    const result = await this.manager.createProviderIngestionRecord(
+      providerName,
+    );
+    if (result) {
+      this.options.logger.info(
+        `incremental-engine: Ingestion record created: '${result.ingestionId}'`,
+      );
+    }
+    return result;
+  }
+
+  async ingestOneBurst(id: string, signal: AbortSignal) {
+    const lastMark = await this.manager.getLastMark(id);
+
+    const cursor = lastMark ? lastMark.cursor : void 0;
```

Comment:
`void 0`? :) could `undefined` work here?

Reply:
It sure could! Done.